### PR TITLE
[feature](merge-cloud) Implement CloudTablet

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by platoneko** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/cloud/cloud_tablet.cpp [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+    if (!st.ok()) {
+        if (st.is<ErrorCode::NOT_FOUND>()) {
+            recycle_cached_data();
+        }
+        return st;
+    }
+
+    if (tablet_meta->tablet_state() != TABLET_RUNNING) [[unlikely]] {
+        // MoW may go to here when load while schema change
+        return Status::Error<INVALID_TABLET_STATE>("invalid tablet state. tablet_id={}",
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:174]: also print state

File: be/src/cloud/cloud_tablet.cpp [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+    return st;
+}
+
+void CloudTablet::add_rowsets(std::vector<RowsetSharedPtr> to_add, bool version_overlap,
+                              bool warmup_delta_data) {
+    if (to_add.empty()) {
+        return;
+    }
+
+    auto add_rowsets_directly = [=, this](std::vector<RowsetSharedPtr>& add_rowsets) {
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:203]: duplicated naming

File: be/src/cloud/cloud_tablet.cpp [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    }
+
+    st = _engine.meta_mgr().sync_tablet_rowsets(this);
+    if (st.is<ErrorCode::NOT_FOUND>()) {
+        recycle_cached_data();
+    }
+    return st;
+}
+
+void CloudTablet::add_rowsets(std::vector<RowsetSharedPtr> to_add, bool version_overlap,
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:197]: add the lock to arg list

File: be/src/cloud/cloud_tablet.cpp [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+            // if start_version <= max_version and end_version > max_version,
+            // we need to add it directly
+            to_add_directly.push_back(to_add_rs);
+            for (auto& [v, rs] : _rs_version_map) {
+                if (to_add_v.contains(v)) {
+                    to_delete.push_back(rs);
+                }
+            }
+            delete_rowsets(to_delete);
+            continue;
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:263]: unnecessary continue

File: be/src/cloud/cloud_tablet.cpp [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+            });
+
+    to_add.erase(remove_it, to_add.end());
+
+    // delete rowsets with overlapped version
+    std::vector<RowsetSharedPtr> to_add_directly;
+    std::vector<RowsetSharedPtr> to_delete;
+    for (auto& to_add_rs : to_add) {
+        // delete rowsets with overlapped version
+        std::vector<RowsetSharedPtr> to_delete;
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:247]: redundant to_delete

File: be/src/cloud/cloud_tablet.cpp [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        } else if (to_add_v.second > _max_version) {
+            // if start_version <= max_version and end_version > max_version,
+            // we need to add it directly
+            to_add_directly.push_back(to_add_rs);
+            for (auto& [v, rs] : _rs_version_map) {
+                if (to_add_v.contains(v)) {
+                    to_delete.push_back(rs);
+                }
+            }
+            delete_rowsets(to_delete);
```
Comment (gavinchou) [inline be/src/cloud/cloud_tablet.cpp:262]: should defer delete_rowses()

File: be/src/cloud/cloud_tablet.cpp (hunk 5 above)
Comment (gavinchou) [reply be/src/cloud/cloud_tablet.cpp:247]: shadow var

**Comment by platoneko** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**